hooks/pre/31-mongo_storage_engine.rb

```diff
@@ -0,0 +1,53 @@
+pulp_db_param = param('katello', 'pulp_db_name') || param('foreman_proxy_content', 'pulp_db_name')
+fail_and_exit('Unknown pulp database name') unless pulp_db_param
+MONGO_DB_NAME = pulp_db_param.value
+
+def fail_and_exit(message)
+  Kafo::Helpers.log_and_say :error, message
+  kafo.class.exit 1
+end
+
+def migration
+  export_dir = '/var/tmp/engine_upgrade'
+  mongo_dir = '/var/lib/mongodb'
+  hiera_file = '/etc/foreman-installer/custom-hiera.yaml'
+  mongo_backup = '/var/tmp/mongo_backup'
+
+  # Create export directory and dump MongoDB
+  logger.info 'Stopping Pulp services except MongoDB'
+  Kafo::Helpers.execute('katello-service stop --exclude rh-mongodb34-mongod')
+  Dir.mkdir_p(export_dir)
+  File.chmod(0700, export_dir)
+  logger.info "Starting mongodump to #{export_dir}"
+  Kafo::Helpers.execute("mongodump --host localhost --out #{export_dir}")
+
+  # Move datafiles out of MongoDB directory so it will start
+  logger.info 'Export done, stopping MongoDB to remove old datafiles'
+  Kafo::Helpers.execute('systemctl stop rh-mongodb34-mongod')
+  logger.info "Moving contents from #{mongo_dir} to #{mongo_backup}"
+  Dir.mkdir_p(mongo_backup)
+  File.chmod(0700, mongo_backup)
+  Kafo::Helpers.execute("mv #{mongo_dir}/* #{mongo_backup}")
+
+  # Start MongoDB and import the dump, fail and notify user of backup if restore does not work.
+  logger.info 'Turning on MongoDB with wiredTiger and starting import'
+  Kafo::Helpers.execute('katello-service start --only rh-mongodb34-mongod')
+  Kafo::Helpers.execute("mongod --storageEngine wiredTiger --dbpath #{mongo_dir}")
```

Comment:
This command actually starts mongo as part of its actions and errors if the line above it starting mongodb exists. From testing, we need to drop line 34.

Reply:
A little more color, this starts a mongodb but not as a daemon, so it just hands:

```
2018-06-14T19:03:54.536+0000 I CONTROL  [initandlisten] MongoDB starting : pid=14590 port=27017 dbpath=/var/lib/mongodb 64-bit host=centos7-katello-3-6.woodford.example.com
2018-06-14T19:03:54.536+0000 I CONTROL  [initandlisten] db version v3.4.9
2018-06-14T19:03:54.537+0000 I CONTROL  [initandlisten] git version: 876ebee8c7dd0e2d992f36a848ff4dc50ee6603e
2018-06-14T19:03:54.537+0000 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.0.2k-fips  26 Jan 2017
2018-06-14T19:03:54.537+0000 I CONTROL  [initandlisten] allocator: tcmalloc
2018-06-14T19:03:54.537+0000 I CONTROL  [initandlisten] modules: none
2018-06-14T19:03:54.537+0000 I CONTROL  [initandlisten] build environment:
2018-06-14T19:03:54.537+0000 I CONTROL  [initandlisten]     distarch: x86_64
2018-06-14T19:03:54.537+0000 I CONTROL  [initandlisten]     target_arch: x86_64
2018-06-14T19:03:54.537+0000 I CONTROL  [initandlisten] options: { storage: { dbPath: "/var/lib/mongodb", engine: "wiredTiger" } }
2018-06-14T19:03:54.568+0000 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=2391M,session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),checkpoint=(wait=60,log_size=2GB),statistics_log=(wait=0),
2018-06-14T19:03:54.666+0000 I CONTROL  [initandlisten] 2018-06-14T19:03:54.666+0000 I CONTROL  [initandlisten] ** WARNING: Access control is not enabled for the database.
2018-06-14T19:03:54.666+0000 I CONTROL  [initandlisten] **          Read and write access to data and configuration is unrestricted.
2018-06-14T19:03:54.666+0000 I CONTROL  [initandlisten] ** WARNING: You are running this process as the root user, which is not recommended.
2018-06-14T19:03:54.666+0000 I CONTROL  [initandlisten] 2018-06-14T19:03:54.667+0000 I CONTROL  [initandlisten] 2018-06-14T19:03:54.667+0000 I CONTROL  [initandlisten] ** WARNING: /sys/kernel/mm/transparent_hugepage/enabled is 'always'.
2018-06-14T19:03:54.667+0000 I CONTROL  [initandlisten] **        We suggest setting it to 'never'
2018-06-14T19:03:54.667+0000 I CONTROL  [initandlisten] 2018-06-14T19:03:54.667+0000 I CONTROL  [initandlisten] ** WARNING: /sys/kernel/mm/transparent_hugepage/defrag is 'always'.
2018-06-14T19:03:54.667+0000 I CONTROL  [initandlisten] **        We suggest setting it to 'never'
2018-06-14T19:03:54.667+0000 I CONTROL  [initandlisten] 2018-06-14T19:03:54.698+0000 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/var/lib/mongodb/diagnostic.data'
2018-06-14T19:03:54.734+0000 I INDEX    [initandlisten] build index on: admin.system.version properties: { v: 2, key: { version: 1 }, name: "incompatible_with_version_32", ns: "admin.system.version" }
2018-06-14T19:03:54.735+0000 I INDEX    [initandlisten]          building index using bulk method; build may temporarily use up to 500 megabytes of RAM
2018-06-14T19:03:54.738+0000 I INDEX    [initandlisten] build index done.  scanned 0 total records. 0 secs
2018-06-14T19:03:54.739+0000 I COMMAND  [initandlisten] setting featureCompatibilityVersion to 3.4
2018-06-14T19:03:54.739+0000 I NETWORK  [thread1] waiting for connections on port 27017
```

Is there a way to create the database without starting the database? Or in a way that it can create it and then shut down to be properly started? Or if we start it, create the database against an already running mongo?

Reply:
There is also `mongorestore` which sounds like a better tool to use.

Reply:
I ended up getting all of this working by removing this step and then:

 1) Before starting mongodb with katello-service, the mongod.conf to set it to wiredTiger instead of mmapv1
 2) Then start mongodb which will create a new wiredTiger database from what I could tell